File: package/ios/Core/Types/CoordinateSystem.swift
//
//  CoordinateSystem.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 17.07.24.
//

import Foundation

@frozen
enum CoordinateSystem: String, JSUnionValue {
  case previewView = "preview-view"
  case camera = "camera"

  init(jsValue: String) throws {
    if let parsed = CoordinateSystem(rawValue: jsValue) {
      self = parsed
    } else {
      throw CameraError.parameter(.invalid(unionName: "hardwareLevel", receivedValue: jsValue))
    }
  }

  var jsValue: String {
    return rawValue
  }
}

File: package/android/src/main/java/com/mrousavy/camera/core/types/CoordinateSystem.kt
package com.mrousavy.camera.core.types

import com.mrousavy.camera.core.InvalidTypeScriptUnionError

enum class CoordinateSystem(override val unionValue: String) : JSUnionValue {
  PREVIEW_VIEW("preview-view"),
  CAMERA("camera");

  companion object : JSUnionValue.Companion<CoordinateSystem> {
    override fun fromUnionValue(unionValue: String?): CoordinateSystem =
      when (unionValue) {
        "preview-view" -> PREVIEW_VIEW
        "camera" -> CAMERA
        else -> throw InvalidTypeScriptUnionError("coordinateSystem", unionValue)
      }
  }
}
